presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java

```diff
@@ -0,0 +1,206 @@
+                    return endOfData();
+                }
+
+                int batchSize = Math.min(selectedPositions.size(), projectBatchSize);
+                Optional<Page> result = processBatch(batchSize);
+
+                // if the page buffer filled up, so halve the batch size and retry
+                if (!result.isPresent()) {
+                    verify(batchSize > 1);
+                    projectBatchSize = projectBatchSize / 2;
+                    continue;
+                }
+
+                Page page = result.get();
+
+                // if we produced a large page, halve the batch size for the next call
+                long pageSize = page.getSizeInBytes();
+                if (page.getPositionCount() > 1 && pageSize > MAX_PAGE_SIZE_IN_BYTES) {
+                    projectBatchSize = projectBatchSize / 2;
+                }
+
+                // if we produced a small page, double the batch size for the next call
+                if (pageSize < MIN_PAGE_SIZE_IN_BYTES && projectBatchSize < MAX_BATCH_SIZE) {
+                    projectBatchSize = projectBatchSize * 2;
+                }
+
+                // remove batch from selectedPositions and previouslyComputedResults
+                selectedPositions = selectedPositions.subRange(batchSize, selectedPositions.size());
+                for (int i = 0; i < previouslyComputedResults.length; i++) {
+                    if (previouslyComputedResults[i] != null && previouslyComputedResults[i].getPositionCount() > batchSize) {
+                        previouslyComputedResults[i] = previouslyComputedResults[i].getRegion(batchSize, previouslyComputedResults[i].getPositionCount() - batchSize);
+                    }
+                    else {
+                        previouslyComputedResults[i] = null;
+                    }
+                }
+
+                return page;
+            }
+        }
+
+        private Optional<Page> processBatch(int batchSize)
+        {
+            Block[] blocks = new Block[projections.size()];
+
+            int pageSize = 0;
+            SelectedPositions positionsBatch = selectedPositions.subRange(0, batchSize);
+            for (int i = 0; i < projections.size(); i++) {
+                if (positionsBatch.size() > 1 && pageSize > MAX_PAGE_SIZE_IN_BYTES) {
+                    return Optional.empty();
+                }
+
+                // if possible, use previouslyComputedResults produced in prior optimistic failure attempt
+                PageProjection projection = projections.get(i);
+                if (previouslyComputedResults[i] != null && previouslyComputedResults[i].getPositionCount() > batchSize) {
+                    blocks[i] = previouslyComputedResults[i].getRegion(0, batchSize);
+                }
+                else {
+                    previouslyComputedResults[i] = projection.project(session, projection.getInputChannels().getInputChannels(page), positionsBatch);
```

Comment:
`projection.getInputChannels().getInputChannels(page)`

Duplicated `getInputChanngls` looks strange. Maybe rename `getInputChannels(page)` to `select`, `cut`, `extract`, `apply`?